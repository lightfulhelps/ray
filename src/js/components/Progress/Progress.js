// @flow
import * as React from 'react';
import classNames from 'classnames';

type Props = {
  children?: React.Node,
  className?: string,
  isAnimated?: boolean,
  isStriped?: boolean,
  theme?: string,
  value: number,
};

const Progress = ({
  children,
  className,
  isAnimated,
  isStriped,
  theme = 'primary',
  value = 0,
  ...other
}: Props) => {
  const classes = classNames(className, 'progress');
  const barClasses = classNames(
    'progress-bar',
    { [`bg-${theme}`]: theme },
    { 'progress-bar-animated': isAnimated },
    { 'progress-bar-striped': isStriped }
  );

  return (
    <div {...other} className={classes} data-test-id="progress">
      <div
        className={barClasses}
        data-test-id="progress-bar"
        role="progressbar"
        style={{ width: `${value}%` }}
        aria-valuenow="25"
        aria-valuemin="0"
        aria-valuemax="100"
      >
        {children}
      </div>
    </div>
  );
};

export default Progress;
