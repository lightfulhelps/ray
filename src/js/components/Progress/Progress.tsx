import * as React from 'react';
import classNames from 'classnames';

type Props = {
  children?: React.ReactNode;
  className?: string;
  isAnimated?: boolean;
  isStriped?: boolean;
  theme?: string;
  value: number;
  size?: 'sm' | 'md' | 'lg';
  [key: string]: any;
};

const Progress: React.FC<Props> = ({
  children,
  className,
  isAnimated,
  isStriped,
  theme = 'primary',
  size = 'md',
  value = 0,
  ...other
}: Props) => {
  const classes = classNames(className, 'progress');
  const barClasses = classNames(
    'progress-bar',
    { [`bg-gradient-${theme}-x bg-${theme}`]: theme && value <= 100 },
    { 'bg-danger': value > 100 },
    { 'progress-bar-animated': isAnimated },
    { 'progress-bar-striped': isStriped }
  );

  return (
    <div
      {...other}
      className={classes}
      data-test-id="progress"
      style={{
        height: `${size === 'sm' ? '0.5rem' : size === 'lg' ? '1.5rem' : '1.25rem'}`,
        fontSize: `${size === 'sm' ? '0.5rem' : size === 'lg' ? '1rem' : '0.75rem'}`,
      }}
    >
      <div
        className={barClasses}
        data-test-id="progress-bar"
        role="progressbar"
        style={{
          width: `${value}%`,
        }}
        aria-valuenow={value}
        aria-valuemin={0}
        aria-valuemax={100}
      >
        {children}
      </div>
    </div>
  );
};

export default Progress;
