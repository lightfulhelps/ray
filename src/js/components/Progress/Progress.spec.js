import React from 'react';
import { shallow } from 'enzyme';
import merge from 'lodash/merge';
import Progress from './Progress';

const setup = (overrides = {}) => {
  const props = merge({}, overrides);
  const wrapper = shallow(<Progress {...props} />);

  return { wrapper, props };
};

describe('<Progress />', () => {
  it('should render', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle className', () => {
    const { wrapper } = setup();

    wrapper.setProps({ className: 'custom' });

    expect(wrapper.hasClass('progress')).toBe(true);
    expect(wrapper.hasClass('custom')).toBe(true);
  });

  it('should handle the isAnimated prop', () => {
    const { wrapper } = setup();

    expect(wrapper.find('[data-test-id="progress-bar"]').hasClass('progress-bar-animated')).toBe(
      false
    );

    wrapper.setProps({ isAnimated: true });

    expect(wrapper.find('[data-test-id="progress-bar"]').hasClass('progress-bar-animated')).toBe(
      true
    );
  });

  it('should handle the isStriped prop', () => {
    const { wrapper } = setup();

    expect(wrapper.find('[data-test-id="progress-bar"]').hasClass('progress-bar-striped')).toBe(
      false
    );

    wrapper.setProps({ isStriped: true });

    expect(wrapper.find('[data-test-id="progress-bar"]').hasClass('progress-bar-striped')).toBe(
      true
    );
  });

  it('should handle the theme prop', () => {
    const { wrapper } = setup();

    expect(wrapper.find('[data-test-id="progress-bar"]').hasClass('bg-primary')).toBe(true);

    wrapper.setProps({ theme: 'secondary' });

    expect(wrapper.find('[data-test-id="progress-bar"]').hasClass('bg-secondary')).toBe(true);
  });

  it('should handle the value prop', () => {
    const { wrapper } = setup();

    expect(wrapper.find('[data-test-id="progress-bar"]').prop('style').width).toBe('0%');

    wrapper.setProps({ value: 30 });

    expect(wrapper.find('[data-test-id="progress-bar"]').prop('style').width).toBe('30%');
  });

  it('should pass through other props', () => {
    const { wrapper } = setup({ tabIndex: 1, id: 'test' });

    expect(wrapper.prop('tabIndex')).toEqual(1);
    expect(wrapper.prop('id')).toEqual('test');
  });
});
