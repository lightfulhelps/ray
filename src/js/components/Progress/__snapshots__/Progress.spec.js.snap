// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Progress /> should render 1`] = `
<div
  className="progress"
  data-test-id="progress"
>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="25"
    className="progress-bar bg-primary"
    data-test-id="progress-bar"
    role="progressbar"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;
